File: bank-account/typescript/tests/AccountBuilder.ts
import { Account } from '../src/Account.js';
import { Money } from '../src/Money.js';
import { FixedClock } from './FixedClock.js';

interface SeededDeposit { date: Date; amount: number; }

export class AccountBuilder {
  private _openedOn: Date = new Date(Date.UTC(2026, 0, 1));
  private _seededDeposits: SeededDeposit[] = [];

  openedOn(date: Date): this { this._openedOn = date; return this; }

  withDepositOn(date: Date, amount: number): this {
    this._seededDeposits.push({ date, amount });
    return this;
  }

  build(): { account: Account; clock: FixedClock } {
    const clock = new FixedClock(this._openedOn);
    const account = new Account(clock);
    for (const { date, amount } of this._seededDeposits) {
      clock.advanceTo(date);
      account.deposit(new Money(amount));
    }
    return { account, clock };
  }
}

File: bank-account/python/tests/account_builder.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import date

from bank_account import Account, Money

from .fixed_clock import FixedClock


@dataclass
class _SeededDeposit:
    when: date
    amount: float


class AccountBuilder:
    def __init__(self) -> None:
        self._opened_on: date = date(2026, 1, 1)
        self._seeded: list[_SeededDeposit] = []

    def opened_on(self, d: date) -> AccountBuilder:
        self._opened_on = d
        return self

    def with_deposit_on(self, d: date, amount: float) -> AccountBuilder:
        self._seeded.append(_SeededDeposit(d, amount))
        return self

    def build(self) -> tuple[Account, FixedClock]:
        clock = FixedClock(self._opened_on)
        account = Account(clock)
        for seed in self._seeded:
            clock.advance_to(seed.when)
            account.deposit(Money(seed.amount))
        return account, clock
